test/e2e/storage/testsuites/volume_io.go

```diff
@@ -0,0 +1,368 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+// MD5 hashes of the test file corresponding to each file size.
+// Test files are generated in testVolumeIO()
+// If test file generation algorithm changes, these must be recomputed.
+var md5hashes = map[int64]string{
+	testpatterns.FileSizeSmall:  "5c34c2813223a7ca05a3c2f38c0d1710",
+	testpatterns.FileSizeMedium: "f2fa202b1ffeedda5f3a58bd1ae81104",
+	testpatterns.FileSizeLarge:  "8d763edc71bd16217664793b5a15e403",
+}
+
+type volumeIOTestSuite struct {
+	name         string
+	testPatterns []testpatterns.TestPattern
+}
+
+var _ TestSuite = &volumeIOTestSuite{}
+
+// InitVolumeIOTestSuite returns volumeIOTestSuite that implements TestSuite interface
+func InitVolumeIOTestSuite() TestSuite {
+	return &volumeIOTestSuite{
+		name: "volumeIO",
+		testPatterns: []testpatterns.TestPattern{
+			testpatterns.DefaultFsInlineVolume,
+			testpatterns.DefaultFsPreprovisionedPV,
+			testpatterns.DefaultFsDynamicPV,
+		},
+	}
+}
+
+func (t *volumeIOTestSuite) getName() string {
+	return t.name
+}
+
+func (t *volumeIOTestSuite) getTestPatterns() []testpatterns.TestPattern {
+	return t.testPatterns
+}
+
+func (t *volumeIOTestSuite) skipUnsupportedTest(tp testpatterns.TestPattern, drv drivers.TestDriver) {
+}
+
+func (t *volumeIOTestSuite) initTestResource() TestResource {
+	return &genericVolumeTestResource{}
+}
+
+func (t *volumeIOTestSuite) createTestInput(tp testpatterns.TestPattern, tr TestResource) interface{} {
+	var (
+		r       *genericVolumeTestResource
+		ok      bool
+		fsGroup *int64
+	)
+
+	if r, ok = tr.(*genericVolumeTestResource); !ok {
+		framework.Failf("Fail to convert testResource to genericVolumeTestResource")
+		return nil
+	}
+
+	drv := r.driver
+	di := drv.GetDriverInfo()
+	f := di.Framework
+	fileSizes := createFileSizes(di.MaxFileSize)
+	volSource := r.volSource
+
+	if volSource == nil {
+		framework.Skipf("Driver %q does not define volumeSource - skipping", di.Name)
+	}
+
+	if di.IsFsGroupSupported {
+		fsGroupVal := int64(1234)
+		fsGroup = &fsGroupVal
+	}
+
+	return volumeIOTestInput{
+		f:         f,
+		name:      di.Name,
+		config:    di.Config,
+		volSource: *volSource,
+		testFile:  fmt.Sprintf("%s_io_test_%s", di.Name, f.Namespace.Name),
+		podSec: v1.PodSecurityContext{
+			FSGroup: fsGroup,
+		},
+		fileSizes: fileSizes,
+	}
+}
+
+func (t *volumeIOTestSuite) execTest(drv drivers.TestDriver, tp testpatterns.TestPattern) {
+	Context(getTestNameStr(t, tp), func() {
+		var (
+			tr  TestResource
+			ti  interface{}
+			vti volumeIOTestInput
+			ok  bool
+		)
+
+		BeforeEach(func() {
+			tr, ti = prepareTest(drv, t, tp)
+			if vti, ok = (ti).(volumeIOTestInput); !ok {
```

Comment:
I like this a lot better than before, but I'm wondering if it's really necessary to do the test input initialization in a common function and have to do this object casting.

Can we just call the common setupResoource here and then create the volumeIOTestInput directly?

Reply:
I understand your intention. I will separate resource initialization logic and testInput creation logic and only make former one to the common function.